main.cpp

```diff
@@ -38,8 +39,275 @@
+    if (decoded.encoding != bech32::Encoding::BECH32) {
+        // Decoding failed, this should only happen when starting from
+        // an empty corpus.
+        return initial_input(fuzz_data, size, max_size);
+    }
+
+    auto data = decoded.data;
+    auto hrp = decoded.hrp;
+
+    size_t data_maxlen = input.size() > 8 ? std::min(input.size() - 8, static_cast<size_t>(65)) : 0;
+    size_t hrp_maxlen = input.size() > 6 ? std::min(input.size() - 6, static_cast<size_t>(83)) : 0;
+
+    // Mutate either the hrp or data
+    std::srand(seed);
+    switch (std::rand() % 2) {
+        case 0: { // Mutate hrp
+            if (hrp_maxlen > 0) {
+                // Make sure we have a buffer that's large enough for mutation
+                std::vector<uint8_t> hrp_buffer(hrp.begin(), hrp.end());
+                // Reserve enough space for the maximum mutation size
+                hrp_buffer.resize(hrp_maxlen);
+
+                // Mutate the buffer
+                size_t new_len = LLVMFuzzerMutate(hrp_buffer.data(),
+                                                 hrp.size(),
+                                                 hrp_maxlen - 1);
+
+                // Ensure minimum length and validate - at least 1 character
+                new_len = std::max(new_len, static_cast<size_t>(1));
+
+                // Convert back to string with the new length
+                hrp = std::string(reinterpret_cast<char*>(hrp_buffer.data()), new_len);
+
+                // Sanitize hrp - ensure only valid ASCII characters (33-126) and not uppercase
+                for (char& c : hrp) {
+                    if (c < 33 || c > 126) {
+                        c = 'a' + (c % 26); // Replace with a valid lowercase letter
+                    } else if (c >= 'A' && c <= 'Z') {
+                        c = c - 'A' + 'a'; // Convert uppercase to lowercase
+                    }
+                }
+            }
+            break;
+        }
+        case 1: { // Mutate data
+            if (data_maxlen > 0) {
+                // Make sure we have a buffer that's large enough for mutation
+                data.resize(std::max(data.size(), data_maxlen));
+
+                size_t datalen = data.size();
+                size_t new_len = LLVMFuzzerMutate(data.data(),
+                                                 std::min(datalen, data_maxlen),
+                                                 data_maxlen);
+
+                new_len = std::max(new_len, static_cast<size_t>(1));
+                data.resize(new_len);
+
+                for (auto& val : data) {
+                    val &= 0x1F;
+                }
+            }
+            break;
+        }
+    }
+
+    if (hrp.empty() || hrp.size() > 83) {
+        return initial_input(fuzz_data, size, max_size);
+    } else if (!hrp.empty()) {
+        for (const char& c : hrp) {
+            if (c >= 'A' && c <= 'Z') {
+                return initial_input(fuzz_data, size, max_size);
+            }
```

Comment:
How do we get in this case, given that we enforce lowercase above?

Reply:
We enforce only if we mutate the hrp part, right? If it takes the path to mutate the data, the HRP is left unchanged, and if it originally contained uppercase letters, those aren't sanitized.